scanner: allow matching archives by member

bkoropoff opened the pull request:
This allows more flexible detection of content packaged as multiple
compressed files (e.g. MAME).

--------------------------

Basically, this lets the scanner detect MAME/FBA content by matching the CRC of a uniquely-identifying file inside an archive.  It's not 100% complete (there is no single unique file to use as a matcher for some games), but it's an effective fallback when matching the CRC of the archive itself doesn't work.  Assuming this is a reasonable approach, I have related patches to libretro-database and libretro-super to enable it.

retro-wertz commented:
what are the chances that the zipped would not match crc(after running
torrentzip as expected) when we have a way to verify file using clrmame and
create the zip which the crc should be mostly identical (that how torrents
are made available for fba/mame). are we planning something like adding
identifier file to existing zip?

On Thu, Nov 9, 2017 at 9:20 AM, Brian Koropoff <notifications@github.com>
wrote:

> This allows more flexible detection of content packaged as multiple
> compressed files (e.g. MAME).
> ------------------------------
>
> Basically, this lets the scanner detect MAME/FBA content by matching the
> CRC of a uniquely-identifying file inside an archive. It's not 100%
> complete (there is no single unique file to use as a matcher for some
> games), but it's an effective fallback when matching the CRC of the archive
> itself doesn't work. Assuming this is a reasonable approach, I have related
> patches to libretro-database and libretro-super to enable it.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/libretro/RetroArch/pull/5663
> Commit Summary
>
>    - scanner: allow matching archives by member
>
> File Changes
>
>    - *M* core_info.c
>    <https://github.com/libretro/RetroArch/pull/5663/files#diff-0> (44)
>    - *M* core_info.h
>    <https://github.com/libretro/RetroArch/pull/5663/files#diff-1> (3)
>    - *M* tasks/task_database.c
>    <https://github.com/libretro/RetroArch/pull/5663/files#diff-2> (11)
>
> Patch Links:
>
>    - https://github.com/libretro/RetroArch/pull/5663.patch
>    - https://github.com/libretro/RetroArch/pull/5663.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/libretro/RetroArch/pull/5663>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWPDtvOltHH_fESDQrMJHaJ5ot5O5kuaks5s0lNVgaJpZM4QXS05>
> .
>

bkoropoff commented:
@retro-wertz As a simple example, someone using 7z instead of zip would get no matches whatsoever.  This also has the advantage of dat files being generated from the freely-distributable XML database of MAME/FBA rather than requiring actual ROMs.  That said, the existing match method is left intact by this change and takes precedence.

retro-wertz commented:
well if this means scanning the way mame/fba does then good..... just
hoping we can use official dats to verify(or update) our romsets

On Thu, Nov 9, 2017 at 9:38 AM, Brian Koropoff <notifications@github.com>
wrote:

> @retro-wertz <https://github.com/retro-wertz> As a simple example,
> someone using 7z instead of zip would get no matches whatsoever. This also
> has the advantage of dat files being generated from the
> freely-distributable XML database of MAME/FBA rather than requiring actual
> ROMs. That said, the existing match method is left intact by this change
> and takes precedence.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libretro/RetroArch/pull/5663#issuecomment-343019776>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWPDtk0H_WVFsucKomS6iAbkOUKdLexAks5s0leygaJpZM4QXS05>
> .
>

twinaphex commented:
OK, does this PR have any known regressions so far? Any downsides?

RobLoach commented:
@bkoropoff What's the best way to test this? I'd be willing to give it a go.... Thanks for the ping, Twinaphex.

bkoropoff commented:
You'll have to build the updated databases from my other PRs and install
them along with the updated core definitions, then try doing a scan.
You'll need to use non-TorrentZipped archives to be sure it's working.
Also try scanning a directory of non-arcade ROMs to check for perf
regressions

On Wed, Nov 15, 2017, 7:49 AM Rob Loach <notifications@github.com> wrote:

> @bkoropoff <https://github.com/bkoropoff> What's the best way to test
> this? I'd be willing to give it a go.... Thanks for the ping, Twinaphex.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/libretro/RetroArch/pull/5663#issuecomment-344634633>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACAQN6vvxIPwauRcmtzl52SL1XMEK0eKks5s2wfrgaJpZM4QXS05>
> .
>

retro-wertz commented:
from the looks of things, the database seems to be for non-split roms, am I right?

bkoropoff commented:
The new data matches individual files that are unique to a particular machine, so it should work with both split and non-merged archives.  It won't work particularly well with merged since multiple machines in the same archive won't result in multiple playlist entries.

retro-wertz commented:
ok, are the files at least from the parent romset? so that when this works, the one added to playlst is the parent rom and not from the incomplete sub/child roms(which will not work anyways if parent rom is incomplete)

twinaphex commented:
OK so what is the consensus on this? Merge / not merge? Does it have any downsides?

bkoropoff commented:
There should be no downsides as it stands now, as far as I know.

On Thu, Nov 16, 2017, 3:20 AM Twinaphex <notifications@github.com> wrote:

> OK so what is the consensus on this? Merge / not merge? Does it have any
> downsides?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libretro/RetroArch/pull/5663#issuecomment-344893993>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACAQN5BHzazi9kf1d3CQ0qKEZaSxAcqZks5s3BqGgaJpZM4QXS05>
> .
>

twinaphex commented:
OK, can you run either ASAN or Dr. Memory before we merge this?

retro-wertz commented:
since this is now merged, did a quick test and i cant get this to detect any of non-trrntzipped split or merged roms. already recompiled retroarch, recompiled database, update core info, and it does not detect from the test_romsets i did(only verified cps1 romsets/fba database, removed incomplete-unfixed entries)

twinaphex commented:
@bkoropoff Got some input on what @retro-wertz is describing?

bkoropoff commented:
@twinaphex @retro-wertz 

I can't reproduce this.  I did the following:

- Unzipped a ROM, re-zipped it, confirmed that the CRC changed
- Copied the new MAME database and core info files into my config dir
- Moved my old playlist out of the way
- Scanned the new ZIP.  Confirmed it scanned the file contents rather hitting a match for the ZIP itself and stopping.
- Confirmed ROM was detected and added to playlist.

retro-wertz commented:
standby, it seems to be scanning ot using default rdb instead of compiled
ones for some reason... ill update how many it detects compared to how many
zip/titles it should be.

On Mon, Nov 20, 2017 at 4:01 PM, Brian Koropoff <notifications@github.com>
wrote:

> @twinaphex <https://github.com/twinaphex> @retro-wertz
> <https://github.com/retro-wertz>
>
> I can't reproduce this. I did the following:
>
>    - Unzipped a ROM, re-zipped it, confirmed that the CRC changed
>    - Copied the new MAME database and core info files into my config dir
>    - Moved my old playlist out of the way
>    - Scanned the new ZIP. Confirmed it scanned the file contents rather
>    hitting a match for the ZIP itself and stopping.
>    - Confirmed ROM was detected and added to playlist.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libretro/RetroArch/pull/5663#issuecomment-345618427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWPDtuW8yZ6harL8Pc5mvFkvGSi1CY-bks5s4THngaJpZM4QXS05>
> .
>

bkoropoff commented:
Note that you *won't* get a 100% detection rate, which would require more
substantial changes to the way RetroArch does scanning.  It's simply better
than missing everything if you don't have specially blessed ZIP files.

On Mon, Nov 20, 2017 at 12:24 AM, retro-wertz <notifications@github.com>
wrote:

> standby, it seems to be scanning ot using default rdb instead of compiled
> ones for some reason... ill update how many it detects compared to how many
> zip/titles it should be.
>
> On Mon, Nov 20, 2017 at 4:01 PM, Brian Koropoff <notifications@github.com>
> wrote:
>
> > @twinaphex <https://github.com/twinaphex> @retro-wertz
> > <https://github.com/retro-wertz>
> >
> > I can't reproduce this. I did the following:
> >
> > - Unzipped a ROM, re-zipped it, confirmed that the CRC changed
> > - Copied the new MAME database and core info files into my config dir
> > - Moved my old playlist out of the way
> > - Scanned the new ZIP. Confirmed it scanned the file contents rather
> > hitting a match for the ZIP itself and stopping.
> > - Confirmed ROM was detected and added to playlist.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/libretro/RetroArch/pull/5663#issuecomment-345618427
> >,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AWPDtuW8yZ6harL8Pc5mvFkvGSi1CY-bks5s4THngaJpZM4QXS05>
> > .
>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libretro/RetroArch/pull/5663#issuecomment-345622995>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACAQN7XveLy4k06zU4g5rTByb3timxmAks5s4TczgaJpZM4QXS05>
> .
>

retro-wertz commented:
here are the scan results. verified with fba dat (not a mame user here)m
and only CPS1, removing incomplete/unfixed ones.

fba merged romsets:
# of zips = 38
# detected entries for torrentzip = 38
# detected entries for non-trrntzip = 38
# missing = 0

fba split romsets:
# of zips = 275
# detected entries for torrentzip = 275
# detected entries for non-trrntzip = 249
# missing = 26


NOTE: i hade to delete all mame*.rdb as it is duplicating/splitting
imported files to both mame.lpl and fbalpha-arcade games.lpl when scanning
non-torrntzipped ones.

On Mon, Nov 20, 2017 at 4:46 PM, Brian Koropoff <notifications@github.com>
wrote:

> Note that you *won't* get a 100% detection rate, which would require more
> substantial changes to the way RetroArch does scanning. It's simply better
> than missing everything if you don't have specially blessed ZIP files.
>
> On Mon, Nov 20, 2017 at 12:24 AM, retro-wertz <notifications@github.com>
> wrote:
>
> > standby, it seems to be scanning ot using default rdb instead of compiled
> > ones for some reason... ill update how many it detects compared to how
> many
> > zip/titles it should be.
> >
> > On Mon, Nov 20, 2017 at 4:01 PM, Brian Koropoff <
> notifications@github.com>
> > wrote:
> >
> > > @twinaphex <https://github.com/twinaphex> @retro-wertz
> > > <https://github.com/retro-wertz>
> > >
> > > I can't reproduce this. I did the following:
> > >
> > > - Unzipped a ROM, re-zipped it, confirmed that the CRC changed
> > > - Copied the new MAME database and core info files into my config dir
> > > - Moved my old playlist out of the way
> > > - Scanned the new ZIP. Confirmed it scanned the file contents rather
> > > hitting a match for the ZIP itself and stopping.
> > > - Confirmed ROM was detected and added to playlist.
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/libretro/RetroArch/pull/5663#
> issuecomment-345618427
> > >,
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > AWPDtuW8yZ6harL8Pc5mvFkvGSi1CY-bks5s4THngaJpZM4QXS05>
> > > .
> >
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/libretro/RetroArch/pull/5663#issuecomment-345622995
> >,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ACAQN7XveLy4k06zU4g5rTByb3timxmAks5s4TczgaJpZM4QXS05>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/libretro/RetroArch/pull/5663#issuecomment-345628194>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWPDttr4YUf7BY7iFgmlAKmQP1akQ4CHks5s4Tx2gaJpZM4QXS05>
> .
>

twinaphex commented:
@bkoropoff NGPC (Neo Geo Pocket Color) No-Intro and Dreamcast games no longer scan.

Are you aware of this and have you been able to reproduce this? It keeps showing the message '0% : Preparing for content scan...' and does not continue from there.

bkoropoff commented:
@twinaphex No, I hadn't heard of it and I can't reproduce it with a recent git master build.
